appserver/ejb/ejb-container/src/main/java/org/glassfish/ejb/deployment/annotation/handlers/AbstractEjbHandler.java

```diff
@@ -233,9 +233,9 @@ public HandlerProcessingResult processAnnotation(AnnotationInfo ainfo)
                                    ejbClass.getName() }));
                 return getDefaultFailedResult();
             }
-
-
-        } else {
+        }
+        // Create a new descriptor only if there are no entries in the ejb-jar with same class name. EJB 3.0 spec
+        else if(currentBundle.getEjbByClassName(ejbClass.getName()).length == 0) {
```

Comment:
No. It will not fall into the if block because the ejbDesc would be null in that case even though we have an entry in the deployment descriptor. This is because of the getEjbByName function call used to get the ejbDesc. 
```ejbDesc = currentBundle.getEjbByName(elementName);```
While handling the annotations, this elementName field is set to the name attribute of the corresponding EJB annotation. In case name is not present, it is set to the default class name.
So, what happens in this particular case is that, since we do not specify a name attribute with the annotation, elementName is set to the class name. When getEjbByName is called with the class name, it returns a null as the ejb DD has an entry with a different class name.
This is why I have added an if check with the getEjbByClassName function call which would return the list of ejbs with same class name. If we find even one such entry in the DD, we do not need to go into the else block.